src/Form/FormProtector.php

```diff
@@ -0,0 +1,605 @@
+        sort($unlocked, SORT_STRING);
+
+        return $unlocked;
+    }
+
+    /**
+     * Generate the token data.
+     *
+     * @return array The token data.
+     * @psalm-return array{fields: string, unlocked: string}
+     */
+    public function buildTokenData(): array
+    {
+        $fields = $this->fields;
+        $unlockedFields = $this->unlockedFields;
+
+        $locked = [];
+        foreach ($fields as $key => $value) {
+            if (is_numeric($value)) {
+                $value = (string)$value;
+            }
+            if (!is_int($key)) {
+                $locked[$key] = $value;
+                unset($fields[$key]);
+            }
+        }
+
+        sort($unlockedFields, SORT_STRING);
+        sort($fields, SORT_STRING);
+        ksort($locked, SORT_STRING);
+        $fields += $locked;
+
+        $fields = $this->generateHash($fields, $unlockedFields, $this->url, $this->sessionId);
+        $locked = implode('|', array_keys($locked));
+
+        return [
+            'fields' => urlencode($fields . ':' . $locked),
+            'unlocked' => urlencode(implode('|', $unlockedFields)),
+            'debug' => urlencode(json_encode([
+                $this->url,
+                $this->fields,
+                $this->unlockedFields,
+            ])),
+        ];
+    }
+
+    /**
+     * Generate validation hash.
+     *
+     * @param array $fields Fields list.
+     * @param array $unlockedFields Unlocked fields.
+     * @param string $url Form URL.
+     * @param string $sessionId Session Id.
+     * @return string
+     */
+    protected function generateHash(array $fields, array $unlockedFields, string $url, string $sessionId)
+    {
+        $hashParts = [
+            $url,
+            serialize($fields),
+            implode('|', $unlockedFields),
+            $sessionId,
+        ];
+
+        return hash_hmac('sha1', implode('', $hashParts), Security::getSalt());
+    }
+
+    /**
+     * Create a message for humans to understand why Security token is not matching
+     *
+     * @param array $formData Data.
+     * @param array $hashParts Elements used to generate the Token hash
+     * @return string Message explaining why the tokens are not matching
+     */
+    protected function debugTokenNotMatching(array $formData, array $hashParts): string
+    {
+        $messages = [];
+        if (!isset($formData['_Token']['debug'])) {
+            return 'Form protection debug token not found.';
+        }
+
+        $expectedParts = json_decode(urldecode($formData['_Token']['debug']), true);
+        if (!is_array($expectedParts) || count($expectedParts) !== 3) {
+            return 'Invalid form protection debug token.';
+        }
+        $expectedUrl = Hash::get($expectedParts, 0);
+        $url = Hash::get($hashParts, 'url');
+        if ($expectedUrl !== $url) {
+            $messages[] = sprintf('URL mismatch in POST data (expected \'%s\' but found \'%s\')', $expectedUrl, $url);
```

Comment:
```suggestion
            $messages[] = sprintf('URL mismatch in POST data (expected `%s` but found `%s`)', $expectedUrl, $url);
```

If you are on-board with using backticks, I can update all the other error messages and fix tests.